Fix defaulter-gen for recursive types

nikhita opened the pull request:
For recursive types, the defaulter currently goes into an infinite recursion.
Fix to support recursive types.

/cc @sttts

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/gengo/61)
<!-- Reviewable:end -->

nikhita commented on examples/defaulter-gen/generators/defaulter.go:
```diff
@@ -320,7 +320,8 @@ func Packages(context *generator.Context, arguments *args.GeneratorArgs) generat
 					continue
 				}
-				if buildCallTreeForType(t, true, existingDefaulters, newDefaulters) != nil {
+				existingTypes := make(map[string]bool)
```
PTAL if this is the right way to do it.

sttts commented on examples/defaulter-gen/generators/defaulter.go:
```diff
@@ -396,7 +397,7 @@ func Packages(context *generator.Context, arguments *args.GeneratorArgs) generat
 // this function skips adding that defaulter - this allows us to avoid generating object defaulter functions for
 // list types that call empty defaulters.
-func buildCallTreeForType(t *types.Type, root bool, existingDefaulters, newDefaulters defaulterFuncMap) *callNode {
+func buildCallTreeForType(t *types.Type, root bool, existingDefaulters, newDefaulters defaulterFuncMap, existingTypes map[string]bool) *callNode {
```
Can you document existingTypes? Is it feasible to have a recursive type in some test? Are there tests here in the generator?

nikhita replied:
Will document it.
Currently there are no tests. But I'll add one.

sttts commented on examples/defaulter-gen/generators/defaulter.go:
```diff
@@ -432,19 +433,26 @@ func buildCallTreeForType(t *types.Type, root bool, existingDefaulters, newDefau
+		return parent
+	}
+	// if type doesn't exist, mark it as existing
+	existingTypes[t.Name.Name] = true
```
shouldn't we check that `Name` is actually files, i.e. not an anonymous type? Otherwise, we stop too early for those, don't we?

nikhita replied:
So I should check for `t.Name`?

sttts replied:
If I read the type godoc correctly, Name can be empty.

sttts commented:
Can you link a commit for the Kubernetes defaulters? Do they change after this PR?

thockin commented:
This could really use output tests like https://github.com/kubernetes/gengo/tree/master/examples/deepcopy-gen/output_tests

nikhita commented:
Updated the defaulter code. PTAL.

I have added some tests. However, I come across the following error when I run `make`: https://gist.github.com/nikhita/379c3789b8049038109b1ea1509b483b. It works when there is no `defaults.go` file though. I have also pushed the file generated in this case.

After making the changes in this PR and fixing the conversion in https://github.com/kubernetes/kubernetes/pull/47263, nothing breaks in Kubernetes (tested against HEAD). This PR shows the cumulative changes in both the defaulter and converter: https://github.com/nikhita/kubernetes/pull/3.

nikhita commented on examples/defaulter-gen/output_tests/builtins/defaults.go:
```diff
@@ -0,0 +1,36 @@
+/*
+Copyright 2016 The Kubernetes Authors.
```
Oops, will fix the year after the review.

nikhita commented:
The tests are being added in https://github.com/kubernetes/gengo/pull/55. Will create a PR to add a test for recursive type on top of https://github.com/kubernetes/gengo/pull/55.

nikhita commented:
Added tests on top of #55 here: https://github.com/caesarxuchao/gengo/pull/1.

sttts commented:
lgtm.

@thockin ptal at https://github.com/caesarxuchao/gengo/pull/1/files for example output.

sttts commented on examples/defaulter-gen/generators/defaulter.go:
```diff
@@ -320,7 +320,10 @@ func Packages(context *generator.Context, arguments *args.GeneratorArgs) generat
-				if buildCallTreeForType(t, true, existingDefaulters, newDefaulters) != nil {
+				// existingTypes keeps track of types that have already been visited in the tree.
+				// This is used to avoid recursion for recursive types.
+				existingTypes := make(map[*types.Type]bool)
```
s/existingTypes/alreadyVisitedTypes/

nikhita replied:
Done.

sttts commented:
@wojtek-t can this go in? (you are in the OWNERs files, together with @lavalamp. Though the merge+review policy here is unclear to me looking at previous PRs).

lavalamp commented on examples/defaulter-gen/generators/defaulter.go:
```diff
@@ -458,20 +469,24 @@ func buildCallTreeForType(t *types.Type, root bool, existingDefaulters, newDefau
+
+	// The type now acts as a parent, not a nested recursive type.
+	// We can now build the tree for it safely.
+	alreadyVisitedTypes[t] = false
```
The name is wrong if you set it to false after you are done. Should be "currentlyBuildingTypes" or "typesInProgress" or something.

lavalamp commented on examples/defaulter-gen/generators/defaulter.go:
```diff
@@ -396,7 +399,7 @@ func Packages(context *generator.Context, arguments *args.GeneratorArgs) generat
 // this function skips adding that defaulter - this allows us to avoid generating object defaulter functions for
 // list types that call empty defaulters.
-func buildCallTreeForType(t *types.Type, root bool, existingDefaulters, newDefaulters defaulterFuncMap) *callNode {
+func buildCallTreeForType(t *types.Type, root bool, existingDefaulters, newDefaulters defaulterFuncMap, alreadyVisitedTypes map[*types.Type]bool) *callNode {
```
Consider making a struct to hold all of the parameters that don't change, and make this function a method on that struct. Will make the code a lot easier to read.

lavalamp commented:
(happy to lgtm & merge after those comments are addressed)

nikhita commented:
@lavalamp Updated.

lavalamp commented on examples/defaulter-gen/generators/defaulter.go:
```diff
@@ -424,32 +447,43 @@ func buildCallTreeForType(t *types.Type, root bool, existingDefaulters, newDefau
 	case types.Pointer:
-		if child := buildCallTreeForType(t.Elem, false, existingDefaulters, newDefaulters); child != nil {
+		r := NewCallTreeForType(c.t.Elem, false, c.existingDefaulters, c.newDefaulters, c.currentlyBuildingTypes)
+		if child := r.buildCallTreeForType(); child != nil {
```
All of these should look like `if child := r.Build(c.t.Elem, false); child != nil {`

lavalamp commented on examples/defaulter-gen/generators/defaulter.go:
```diff
@@ -396,22 +419,22 @@ func Packages(context *generator.Context, arguments *args.GeneratorArgs) generat
 // this function skips adding that defaulter - this allows us to avoid generating object defaulter functions for
 // list types that call empty defaulters.
-func buildCallTreeForType(t *types.Type, root bool, existingDefaulters, newDefaulters defaulterFuncMap) *callNode {
+func (c callTreeForType) buildCallTreeForType() *callNode {
```
`func (c *callTreeForType) Build(t *types.Type, root bool) *callNode {`

lavalamp commented on examples/defaulter-gen/generators/defaulter.go:
```diff
@@ -320,7 +320,11 @@ func Packages(context *generator.Context, arguments *args.GeneratorArgs) generat
+				// This is used to avoid recursion for recursive types.
+				currentlyBuildingTypes := make(map[*types.Type]bool)
+				c := NewCallTreeForType(t, true, existingDefaulters, newDefaulters, currentlyBuildingTypes)
+				if c.buildCallTreeForType() != nil {
```
This should be a one-liner:

```
if NewCallTreeForType(existingDefaulters, newDefaulters).Build(t, true) != nil {
```
 

lavalamp commented on examples/defaulter-gen/generators/defaulter.go:
```diff
@@ -387,6 +391,25 @@ func Packages(context *generator.Context, arguments *args.GeneratorArgs) generat
+	currentlyBuildingTypes map[*types.Type]bool
+}
+
+func NewCallTreeForType(t *types.Type, root bool, existingDefaulters, newDefaulters defaulterFuncMap, currentlyBuildingTypes map[*types.Type]bool) callTreeForType {
```
Actually both this and Build should be private.

nikhita replied:
We can't initialize currentlyBuildingTypes in newCallTreeForType because when this is encountered:

```
r := newCallTreeForType(c.existingDefaulters, c.newDefaulters)
	if child := r.build(t.Elem, false); child != nil
```

it will create a new map for currentlyBuildingTypes - destroying our purpose for storing the types that are currently building. :)

nikhita commented:
@lavalamp Updated. PTAL, thanks.

lavalamp commented on examples/defaulter-gen/generators/defaulter.go:
```diff
@@ -432,19 +450,30 @@ func buildCallTreeForType(t *types.Type, root bool, existingDefaulters, newDefau
 	switch t.Kind {
 	case types.Pointer:
-		if child := buildCallTreeForType(t.Elem, false, existingDefaulters, newDefaulters); child != nil {
+		r := newCallTreeForType(c.existingDefaulters, c.newDefaulters, c.currentlyBuildingTypes)
```
What I was trying to communicate is that you *don't* call newCallTreeForType here, you reuse `c`.

nikhita replied:
Right, sorry - should have done this in the first place.

Updated. Please check if this is better.

lavalamp commented on examples/defaulter-gen/generators/defaulter.go:
```diff
@@ -458,20 +481,24 @@ func buildCallTreeForType(t *types.Type, root bool, existingDefaulters, newDefau
+
+	// The type now acts as a parent, not a nested recursive type.
+	// We can now build the tree for it safely.
+	c.currentlyBuildingTypes[t] = false
```
Looking good--last question, should this be in a defer block? The return on line 496 skips it, for example.

nikhita replied:
Yes. Done.

lavalamp commented:
LGTM, thanks for bearing with me :)
